### PR TITLE
ARROW-3888: [C++] Fix various compiler warnings

### PR DESCRIPTION
**Pull request by fsaintjacques**
- Use of uninitialized variables
- Strict aliasing

### CODE REVIEW

**Comment by wesm** [comment]
Can we just set `-fno-strict-aliasing`?

**Comment by wesm** [comment]
Another matter on which I must plead ignorance =) I frequently see -fno-strict-aliasing in other build systems, and I'm curious to learn more about the practical consequences of violating strict aliasing rules (and why some projects choose to suppress these checks)

**Comment by fsaintjacques** [comment]
TLDR: Breaking strict aliasing can/will induce undefined behaviour. You're then at the mercy of newer compilers deciding if `-O3` will break your code.

The usual fix is to just use memcpy, the compiler will hopefully elide the call if it can deduce the size parameter to be constant and smallish.

As for using `-fno-strict-aliasing`, sometimes fixing the strict aliasing breakage is too cumbersome (due to huge codebase or lack of control of dependencies).

In our case, there was only one real reported warning and various usage in tests. 

Take a short read  https://blog.regehr.org/archives/959

**Review by wesm — APPROVED** [review]
+1, thanks for the link!
